### ISSUE TITLE
about train

### ISSUE THREAD

**Issue by ziziyaziziya** [FreeformRobotics/OTS#1]
hello ~~. How to train your model ?

**Comment by bo-miao** [comment]
Hi,

We compute the object features offline and store them on the server, e.g. *.json, *.npy, to accelerate the training speed. After that, we train the model based on the pre-computed object features of the images in the training set with cross-entropy loss and SGD optimizer.

Below is the architecture of the dataset.
```Shell
├── dataset
    ├── train
        ├── class names
            ├── images
    ├── valid
        ├── class names
            ├── images
```

**Comment by ziziyaziziya** [comment]
Thanks～～.    I have one more question, how to get segmentation results??? I ran the test code, but only got scene information, no segmentation results...

**Comment by bo-miao** [comment]
You need to modify the code of `self.decoder` of the `self.seg_model` according to the code of PSPNet (https://github.com/CSAILVision/semantic-segmentation-pytorch/blob/master/mit_semseg/models/models.py).

**Comment by pranay-ar** [comment]
Hey @bo-miao, could you also kindly provide the object features file? It'd be very helpful for me.

**Comment by bo-miao** [comment]
> 

Hi @pranay-ar , 

I am sorry that I did not bring the object features with me after I left CUHKSZ since the features are cumbersome (plenty of GB) and I have switched to the segmentation task. 
The extraction of object features **exactly** follows the function "get_object_feature" in "main.py". You can easily modify that code for training data to extract object features. Moreover, the code can be optimized to be much faster. 
BTW, previously I have tried to merge OTS with other types of features, and the performance can be further improved by more than 1%.

**Comment by bo-miao** [comment]
> Hey @bo-miao, could you also kindly provide the object features file? It'd be very helpful for me.

Hi @pranay-ar ,

I am sorry that I did not bring the object features with me after I left CUHKSZ since the features are cumbersome (plenty of GB) and I have switched to the segmentation task.
The extraction of object features exactly follows the function "get_object_feature" in "main.py". You can easily modify that code for training data to extract object features. Moreover, the code can be optimized to be much faster.
BTW, previously I have tried to merge OTS with other types of features, and the performance can be further improved by more than 1%.

**Comment by pranay-ar** [comment]
Hello @bo-miao , sure that is not a problem at all. If possible, could you kindly share the Places365-7 or Places365-14 datasets that you have used in your experiments? Even just mentioning the dataset structure would be sufficient. 

My main intention of using your code is to extract object features individually (from all the classes detected in the scene from  the segmentation head) and save them in a JSON file instead of aggregating them using the OFAM module. Do you think that is something possible to do from your code? I looked at your code and felt that is something doable but I wanted our opinion on it (if you had considered extracting object features individually in your experiments before).

**Comment by bo-miao** [comment]
> Hello @bo-miao , sure that is not a problem at all. If possible, could you kindly share the Places365-7 or Places365-14 datasets that you have used in your experiments? Even just mentioning the dataset structure would be sufficient.
> 
> My main intention of using your code is to extract object features individually (from all the classes detected in the scene from the segmentation head) and save them in a JSON file instead of aggregating them using the OFAM module. Do you think that is something possible to do from your code? I looked at your code and felt that is something doable but I wanted our opinion on it (if you had considered extracting object features individually in your experiments before).

Places365-7 and Places365-14 are subsets of Places365, you can refer to Places365 for the data structure. You can find the class names in IV. EXPERIMENTS AND RESULTS (B.). Just extract the corresponding class data from Places365 to form Places365-7 and Places365-14.

Yes, of course, you can extract object features with our code since that is exactly what we did before OFAM. The code could be refined to be faster or lighter since we did not spend too much effort on that. I am not sure if JSON is a suitable way to store large embedding features. You can try it or try other formats such as .npy.

**Comment by pranay-ar** [comment]
Hi @bo-miao, thank you for the help. I believe I was able to get the job done. I just wanted to confirm one last thing. In the main.py file at [line 103](https://github.com/FreeformRobotics/OTS/blob/cbe398564c6f322f2e89e430ead24fe7d0f1f81c/main.py#L103), this is the matrix (object_feature matrix that is of size (150,1024)) that contains the feature vectors for each of the 150 objects right? If I have to access a feature vector of an object from the 150 objects, how do I efficiently index it w.r.t this matrix? Is the ordering of 150 objects in the (150,1024) similar to the order that is mentioned in the [object150_info.csv ](https://github.com/FreeformRobotics/OTS/blob/main/data/object150_info.csv) file? Or is the ordering of objects in a chronological manner? I have assumed the former. Kindly let me know if that is the correct order. Thank you!

**Comment by bo-miao** [comment]
> Hi @bo-miao, thank you for the help. I believe I was able to get the job done. I just wanted to confirm one last thing. In the main.py file at [line 103](https://github.com/FreeformRobotics/OTS/blob/cbe398564c6f322f2e89e430ead24fe7d0f1f81c/main.py#L103), this is the matrix (object_feature matrix that is of size (150,1024)) that contains the feature vectors for each of the 150 objects right? If I have to access a feature vector of an object from the 150 objects, how do I efficiently index it w.r.t this matrix? Is the ordering of 150 objects in the (150,1024) similar to the order that is mentioned in the [object150_info.csv ](https://github.com/FreeformRobotics/OTS/blob/main/data/object150_info.csv) file? Or is the ordering of objects in a chronological manner? I have assumed the former. Kindly let me know if that is the correct order. Thank you!

The order of 150x1024 object features is consistent with the .csv file, i.e., the first 1x1024 refers to the first object class in .csv.

**Comment by pranay-ar** [comment]
Thank you very much for your help and for the great work. Apologies for troubling you with so many questions 😅

**Comment by bo-miao** [comment]
> Thank you very much for your help and for the great work. Apologies for troubling you with so many questions sweat_smile

No worries. Hope this work could help with your approach.
